tests/suite-integration/PluginFlyvemdmAgent.php

```diff
@@ -92,144 +96,118 @@ public function testDeviceCountLimit() {
+         ],
+         'with a too low version' => [
+            'data'     => [
+               'version'     => '1.9.0',
+            ],
+            'expected' => 'The agent version is too low',
+         ],
+         'without serial or uuid' => [
+            'data'     => [
+               'serial'      => null,
+            ],
+            'expected' => 'One of serial and uuid is mandatory',
+         ],
+         'without inventory'      => [
+            'data'     => [
+               'version'     => $version,
+               'inventory'   => '',
+            ],
+            'expected' => 'Device inventory XML is mandatory',
+         ],
+         'with invalid inventory' => [
+            'data'     => [
+               'serial'      => $serial,
+               'version'     => $version,
+               'inventory'   => $inventory,
+            ],
+            'expected' => 'Inventory XML is not well formed',
+         ],
+      ];
+   }
+
    /**
-    * @tags testEnrollAgent
+    * @dataProvider invalidEnrollmentDataProvider
+    * @tags testInvalidEnrollAgent
+    * @param array $data
+    * @param string $expected
     */
-   public function testEnrollAgent() {
-      global $DB;
-
-      // Set a computer type
-      $computerTypeId = 3;
-      \Config::setConfigurationValues('flyvemdm', ['computertypes_id' => $computerTypeId]);
-      $expectedLogCount = countElementsInTable(\PluginFlyvemdmInvitationlog::getTable());
-
+   public function testInvalidEnrollAgent(array $data, $expected) {
       list($user, $serial, $guestEmail, $invitation) = $this->createUserInvitation(\User::getForeignKeyField());
+      $invitationToken = (isset($data['invitationToken'])) ? $data['invitationToken'] : $invitation->getField('invitation_token');
+      $serial = (key_exists('serial', $data)) ? $data['serial'] : $serial;
+      $mdmType = (key_exists('mdmType', $data)) ? $data['mdmType'] : 'android';
+      $version = (key_exists('version', $data)) ? $data['version'] : '';
+      $inventory = (key_exists('inventory', $data)) ? $data['inventory'] : null;
 
-      $invitationToken = $invitation->getField('invitation_token');
-      $inviationId = $invitation->getID();
-
-      // Test enrollment with bad token
-      $_SESSION['MESSAGE_AFTER_REDIRECT'] = [];
-      $agent = $this->agentFromInvitation($user, $guestEmail, $serial, 'bad token');
-      $this->boolean($agent->isNewItem())
-         ->isTrue(json_encode($_SESSION['MESSAGE_AFTER_REDIRECT'], JSON_PRETTY_PRINT));
-      $this->array($_SESSION['MESSAGE_AFTER_REDIRECT'][ERROR])->contains('Invitation token invalid');
-
-      // Test the invitation log did not increased
-      // this happens because the enrollment failed without identifying the invitation
-      $invitationLog = new \PluginFlyvemdmInvitationlog();
-      $rows = $invitationLog->find("`plugin_flyvemdm_invitations_id` = '$inviationId'");
-      $this->integer(count($rows))->isEqualTo($expectedLogCount);
```

Comment:
We loose the tests on invitation logs. Can you re-add it while keeping the factorization ?

Reply:
I updated this but the logic has changed a little bit

Reply:
No problem  !